assets/settings.txt
// The currently selected theme
// Value must be a valid filename in `assets/themes` (e.g. `blue` for `blue.txt`)
Theme: blue

// The interpolation speed for camera panning
// Lower is slower (0.0 - No movement, 1.0 - Instant)
CameraPanSpeed: 0.4

// The interpolation speed for camera zooming
// Lower is slower (0.0 - No movement, 1.0 - Instant)
CameraZoomSpeed: 0.4

// The interpolation speed for scrolling
// Lower is slower (0.0 - No movement, 1.0 - Instant)
PopupScrollSpeed: 0.4

// Renders connections as bezier curves or linear lines
// Valid values: `bezier`, `linear`
ConnectionType: bezier

// Switches camera movement to the original world editor's control scheme
// FloodForge: Middle click to pan, left click to select
// World Editor: Left click on empty space to pan, Shift+left click to select
OriginalControls: false

// If true, icons in the World Editor keep a constistant size when zooming
SelectorScale: true

// The default path to start file popups at
// If specified, the first time opening a file popup it will attempt to load this path.
// If the path does not exist, falls back to normal
DefaultFilePath: 

// If true, file popups retain their path upon closing
// E.g. If you import a region from `mods/moreslugcats/world/cl/`, the next time you open a file popup, it will start at that path
KeepFilesystemPath: true

// If true, FloodForge gives a warning about missing `XX_A01_1.png` images
// Note: Makes world loading slower
WarnMissingImages: false

// If false, the tutorial popup appears after closing the splash
HideTutorial: false

// If true, exporting will modify the existing `world_xx.txt` file
// If false, exported worlds are stored in `FloodForge/worlds` and need to be manually copied into their respective directories
UpdateWorldFiles: true

// The default color for rooms without a subregion when `Subregion Colors` is enabled
// Format: #RRGGBB
NoSubregionColor: #ffffff

// List of default colors for rooms when `Subregion Colors` is enabled
// The color picked is based on subregion index, first color for first subregion, second for second, e.t.c. Wraps when out of colors
// Format: #R1G1B1, #R2G2B2, ...
SubregionColors: #ff0000, #00ff00, #0000ff, #ffff00, #00ffff, #ff00ff, #ff7f00, #7f7f7f, #7f00ff, #ff7fff

// How strong of a color tinting is applied to rooms when either `Layer Colors` or `Subregion Colors` is enabled.
// Recommended range: 0.5 - 1.0
RoomTintStrength: 0.5


// DEBUG ONLY!
// If true, exported `map_xx.png` files will have a cyan 10px border instead of lime
DebugVisibleOutputPadding: false

src/Settings.cpp
#include "Settings.hpp"

#include "Logger.hpp"

std::unordered_map<Settings::Setting, SettingType> Settings::settings;

void Settings::loadDefaults() {
	settings[Setting::CameraPanSpeed] = 0.4;
	settings[Setting::CameraZoomSpeed] = 0.4;
	settings[Setting::PopupScrollSpeed] = 0.4;
	settings[Setting::ConnectionType] = 0;
	settings[Setting::OrignalControls] = false;
	settings[Setting::SelectorScale] = true;
	settings[Setting::DefaultFilePath] = "NON_EXISTANT_PATH_YOU_CAN'T_HAVE_THIS_PATH_PLSPLSPLS///";
	settings[Setting::WarnMissingImages] = false;
	settings[Setting::HideTutorial] = false;
	settings[Setting::KeepFilesystemPath] = true;
	settings[Setting::UpdateWorldFiles] = true;
	settings[Setting::DebugVisibleOutputPadding] = false;
	settings[Setting::NoSubregionColor] = Colour(1.0, 1.0, 1.0);
	settings[Setting::RoomTintStrength] = 0.5;

	std::vector<Colour> subregionColors;
	subregionColors.push_back(Colour(1.0, 0.0, 0.0)); // #ff0000
	subregionColors.push_back(Colour(0.0, 1.0, 0.0)); // #00ff00
	subregionColors.push_back(Colour(0.0, 0.0, 1.0)); // #0000ff
	subregionColors.push_back(Colour(1.0, 1.0, 0.0)); // #ffff00
	subregionColors.push_back(Colour(0.0, 1.0, 1.0)); // #00ffff
	subregionColors.push_back(Colour(1.0, 0.0, 1.0)); // #ff00ff
	subregionColors.push_back(Colour(1.0, 0.5, 0.0)); // #ff7f00
	subregionColors.push_back(Colour(0.5, 0.5, 0.5)); // #7f7f7f
	subregionColors.push_back(Colour(0.5, 0.0, 1.0)); // #7f00ff
	subregionColors.push_back(Colour(1.0, 0.5, 1.0)); // #ff7fff
	settings[Setting::SubregionColors] = subregionColors;
}

void Settings::init() {
	loadDefaults();

	std::filesystem::path settingsPath = BASE_PATH / "assets" / "settings.txt";
	if (!std::filesystem::exists(settingsPath)) return;

	std::fstream settingsFile(settingsPath);

	std::string line;
	while (std::getline(settingsFile, line)) {
		if (line.empty()) continue;
		if (line.back() == '\r') line.pop_back();
		if (startsWith(line, "//")) continue;

		std::string key = line.substr(0, line.find_first_of(':'));
		std::string value = line.substr(line.find_first_of(':') + 2);

		try {
			bool boolValue = (toLower(value) == "true" || toLower(value) == "yes" || toLower(value) == "1");
			
			if (key == "Theme") loadTheme(value);
			else if (key == "CameraPanSpeed") settings[Setting::CameraPanSpeed] = std::stod(value);
			else if (key == "CameraZoomSpeed") settings[Setting::CameraZoomSpeed] = std::stod(value);
			else if (key == "PopupScrollSpeed") settings[Setting::PopupScrollSpeed] = std::stod(value);
			else if (key == "ConnectionType") settings[Setting::ConnectionType] = int(toLower(value) == "bezier");
			else if (key == "OriginalControls") settings[Setting::OrignalControls] = boolValue;
			else if (key == "SelectorScale") settings[Setting::SelectorScale] = boolValue;
			else if (key == "DefaultFilePath") settings[Setting::DefaultFilePath] = value;
			else if (key == "WarnMissingImages") settings[Setting::WarnMissingImages] = boolValue;
			else if (key == "HideTutorial") settings[Setting::HideTutorial] = boolValue;
			else if (key == "KeepFilesystemPath") settings[Setting::KeepFilesystemPath] = boolValue;
			else if (key == "UpdateWorldFiles") settings[Setting::UpdateWorldFiles] = boolValue;
			else if (key == "DebugVisibleOutputPadding") settings[Setting::DebugVisibleOutputPadding] = boolValue;
			else if (key == "NoSubregionColor") settings[Setting::NoSubregionColor] = stringToColour(value);
			else if (key == "RoomTintStrength") settings[Setting::RoomTintStrength] = std::stod(value);
			else if (key == "SubregionColors") {
				std::vector<Colour> subregionColors;
				for (std::string item : split(value, ", ")) {
					subregionColors.push_back(stringToColour(item));
				}
				settings[Setting::SubregionColors] = subregionColors;
			};
		} catch (...) {
			Logger::error("Error while loading setting: ", key, " - ", value);
		}
	}

	settingsFile.close();
}

void Settings::cleanup() {

}